### PR TITLE
Add Resource effect and port to fused-effects 1.0.

### PR DESCRIPTION
**Pull request by patrickt**
After spirited discussion, @robrix and I decided that the most appropriate place for the `Resource` effect and carrier is in this separate package, since it violates the state effect laws if used in conjunction with the traditional state carrier.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+install:
+  - cabal new-update
+  - cabal new-build --enable-tests --enable-benchmarks
+
+script:
+  - cabal new-test --enable-tests
+
+notifications:
+  email: false
```
Comment (robrix) [inline .travis.yml]: Can we use the `.travis.yml` file from `fused-effects`? We put some work into optimizing it, as I recall.

File: src/Control/Effect/Catch.hs [hunk 2]
```diff
@@ -13,18 +13,17 @@ module Control.Effect.Catch
...
+import           Control.Carrier
 import           Control.Effect.Reader
 import           Control.Effect.Sum
 import qualified Control.Exception as Exc
 import           Control.Exception.Safe (isSyncException)
 import           Control.Monad.IO.Class
 import           Control.Monad.IO.Unlift
 
+-- | @since 0.1.0.0
 data Catch m k
```
Comment (robrix) [inline src/Control/Effect/Catch.hs:27]: I kinda feel like we should rename this to `CatchIO`. Thoughts?

**Review by robrix — APPROVED** [review]

**Comment by patrickt** [comment]
Gonna pull the .travis.yml thing into a separate PR since this won’t build successfully until we drop 1.0 anyway.

File: src/Control/Carrier/Catch.hs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE FlexibleInstances, GeneralizedNewtypeDeriving, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+
+-- | A carrier for a 'Catch' effect.
+module Control.Carrier.Catch
```
Comment (robrix) [inline src/Control/Carrier/Catch.hs:4]: (If we rename the `Catch` effect to `CatchIO` we should mirror that here as well.)

File: src/Control/Effect/Catch.hs (hunk 2 above)
Comment (patrickt) [reply src/Control/Effect/Catch.hs:27]: Ehh, I could go either way. Let’s think about it. I’m gonna merge this in the meantime.

File: src/Control/Carrier/Catch.hs (hunk 3 above)
Comment (patrickt) [reply src/Control/Carrier/Catch.hs:4]: Or maybe Control.Carrier.Exception.Catch ?

Comment (robrix) [reply src/Control/Carrier/Catch.hs:4]: I could live with that if we rename the effect to `Exception`, totally.

**Pull request merged**